### PR TITLE
rework batch events

### LINKED ISSUE #12095: Refactor batched triggered abilities to expose which part of batch event made them trigger.
So with [[Felix Five Boots]] in #12074, we have an issue of Felix incrementing the trigger count for batch events, where the part of the batch Felix cares about has no common part with the part the batch trigger cares about.

I suspect this might occur in a variety of situations, but for now the one I can illustrate the problem with is the following:
- PlayerA has Felix, an Elite Vanguard equipped with [[Umezawa's Jitte]].
- PlayerA attacks with its 2 creatures.
- PlayerB blocks the creature with Jitte, but the other is not blocked.
- A `DAMAGED_BATCH_FOR_ALL `is triggered with all creatures dealing damage. There is two events in it: Felix dealing damage to playerB, and Elite Vanguard dealing damage to the blocker (there is also the playerB's blocker damage, but irrelevant here)
- Jitte's trigger is checking the trigger -> there is the creature dealing damage in it. Jitte's trigger does trigger on the batched event.
- Felix sees the Jitte's triggering with the `DAMAGED_BATCH_FOR_ALL`, there is the event for Elite Vanguard's damage in it, so it is increased by Felix. This is where the problem is. 

So to solve the problem, I think we need to improve TriggeredEvents to expose which part of events are relevant to them when they do trigger. For non-batched events, this is trivial, if it did trigger, then the event is what made the trigger triggers.

For Batch events, it is a on a case-by-case basis. For instance Jitte is using `DealsCombatDamageEquippedTriggeredAbility`:
image

Here the part of the batch that are relevant are quite explicit:
```java
Permanent sourcePermanent = getSourcePermanentOrLKI(game);
if (sourcePermanent == null || sourcePermanent.getAttachedTo() == null) {
    return false;
}
((DamagedBatchAllEvent) event)
                .getEvents()
                .stream()
                .filter(DamagedEvent::isCombatDamage)
                .filter(e -> e.getAttackerId().equals(sourcePermanent.getAttachedTo()))
```
                
So I would be in favor of separating checkTrigger in two:
```java
@Override
public boolean checkEventType(GameEvent event, Game game) {
    return event.getType() == GameEvent.EventType.DAMAGED_BATCH_FOR_ALL;
}

@Override
public Stream<GameEvent> filterEvent(GameEvent event, Game game) {
    Permanent sourcePermanent = getSourcePermanentOrLKI(game);
    if (sourcePermanent == null || sourcePermanent.getAttachedTo() == null) {
        return Stream.empty();
    }
    return ((DamagedBatchAllEvent) event)
            .getEvents()
            .stream()
            .filter(DamagedEvent::isCombatDamage)
            .filter(e -> e.getAttackerId().equals(sourcePermanent.getAttachedTo()))
}

@Override
public boolean checkTrigger(GameEvent event, Game game) {
    int amount = filterEvent(event, game)
            .mapToInt(GameEvent::getAmount)
            .sum();
    if (amount < 1) {
        return false;
    }
    this.getEffects().setValue("damage", amount);
    return true;
}
```

I think that would help us going forward, as we introduce more batch events and new cards with various `one or more` triggers are more and more frequent.

To go back on Felix, the filtering for Felix would call filterEvent of the original trigger to have the relevant part of the batch, then do its own filtering to decide if it triggers.

### LINKED ISSUE #12193: Kaya Spirits' Justice
Morning,
Was playtesting.  I think based on my read that Kaya's static ability should only allow me to choose from my own creatures.  However, in using her -2 it is allowing me to choose anything I exiled of my opponents.
image

### LINKED ISSUE #13057: Individual triggers on mass damage events can't be handled correctly
There are a few permanents that trigger whenever another permanent deals damage. These include [Justice](https://scryfall.com/card/5ed/41/justice), [Noble Purpose](https://scryfall.com/card/8ed/31/noble-purpose), and new card in FDN [Quilled Greatwurm](https://scryfall.com/card/fdn/111/quilled-greatwurm).

Currently, these effects are implemented by checking a GameEvent and seeing if it is an instance of either `GameEvent.EventType.DAMAGED_PERMANENT` or `GameEvent.EventType.DAMAGED_PLAYER`. However, this means that when a creature deals damage to two permanents/players, the ability will trigger twice, when it should only trigger once.

Replicate with the following `init.txt`:
```
battlefield:Human:Noble Purpose:1
battlefield:Human:Grand Melee:1
battlefield:Human:Hill Giant:1
battlefield:Computer:Gray Ogre:2
```

In this scenario, Human attacks with Hill Giant, it's double blocked, and it deals 2 damage to the first Gray Ogre and 1 to the second. Noble Purpose should only trigger once and gain 3 life, but instead it triggers twice to gain 2 and 1 life in succession.

This is a pretty minor issue but it's a bit more noticeable with the new card Quilled Greatwurm, where players might care about the correct behavior. One example is abilities that care about the number of +1/+1 counters being added to permanents. An example is [Winding Constrictor](https://scryfall.com/card/aer/140/winding-constrictor) which will add too many counters to creatures that deal damage, if the current implementation is used.

I tried working around this issue using `GameEvent.EventType.DAMAGED_BATCH_FOR_ALL`, but I couldn't get it to work. Maybe I haven't dug deep enough though, so let me know if there's a better workaround!

### PR DESCRIPTION
**Pull request by xenohedron**
This is the updated version of #12173.

Many new common classes for batch triggered abilities, which now use a more modular framework. Removed a variety of ad-hoc workarounds, deduplicated some code, and fixed various edge cases.

~~Before merge, I'll drop the hacky verify check. I'll also drop the tests that require test API updates; those can be a separate scope, but I did want to include just in case. (edit:done)~~

- [x] Implement framework
- [x] Tapped, untapped batch events
- [x] Sacrifice batch events
- [x] Milled batch events
- [x] Ob Nixilis, Captive Kingpin
- [x] Zone change batch events (enters battlefield)
- [x] Zone change batch events (not battlefield)
- [x] Zone change batch events (leaves battlefield)
- [x] Fix old #12193
- [x] Damage batch events
- [x] Add new damage batch by source for #13057 
- [x] Add tests from Susucr
- [x] Fix original problem #12095

Separate scope for followup: - Add new life lost by player from #12202

Fix #12193, fix #12095, resolves #13057, closes #12173

### CODE REVIEW

File: Mage.Verify/src/test/java/mage/verify/VerifyCardDataTest.java [hunk 1]
```diff
@@ -1852,6 +1856,34 @@ private void compareClassRecursive(Object obj1, Object obj2, Card originalCard,
...
+                .of(GameEvent.EventType.values())
+                .filter(GameEvent.EventType::isBatch)
+                .map(eventType -> new GameEvent(eventType, null, null, null))
+                .collect(Collectors.toSet());
+    }
+
+    /**
+     * Perform checks on abilities
+     */
+    private void checkAbility(Card originalCard, Ability ability, String msg) {
```
Comment (JayDi85) [inline Mage.Verify/src/test/java/mage/verify/VerifyCardDataTest.java:1873]: Must remove such code after all fixes. It’s too hacky.

Comment (xenohedron) [reply Mage.Verify/src/test/java/mage/verify/VerifyCardDataTest.java]: Done
